### PR TITLE
Project wizard app representation running create client [wizard]

### PR DESCRIPTION
**Pull request by witmicko**
**Description**
Use Project Wizard to get data for `mobile create client` command, and execute it.
Upon execute we are taken to the new pane and the output is shown.

- also added `CodeStyle.xml` - code editor formatter settings from Intelij Idea. I believe we need to agree on the same style and start using it.

**This PR fixes the following issue**
[AEROGEAR-1928](https://issues.jboss.org/browse/AEROGEAR-1928)

jan-31-2018 15-57-29

### CODE REVIEW

File: gradle/wrapper/gradle-wrapper.properties [hunk 1]
```diff
@@ -1,6 +0,0 @@
-#Fri Jan 12 14:51:45 GMT 2018
-distributionBase=GRADLE_USER_HOME
-distributionPath=wrapper/dists
-zipStoreBase=GRADLE_USER_HOME
-zipStorePath=wrapper/dists
-distributionUrl=https\://services.gradle.org/distributions/gradle-3.5-rc-2-all.zip
```
Comment (secondsun) [inline gradle/wrapper/gradle-wrapper.properties]: Why is this file deleted?  It is necessary for the gradle wrapper to start up.

File: src/main/java/org/aerogear/plugin/intellij/mobile/module/AppRepresentationModuleBuilder.java [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+    return new AppRepresentationModuleType();
+  }
+
+  @Override
+  public ModuleWizardStep[] createWizardSteps(WizardContext wizardContext, ModulesProvider modulesProvider) {
+    return new ModuleWizardStep[]{new AppRepresentationWizardStep(this)};
+  }
+
+  @Override
+  public void moduleCreated(@NotNull Module module) {
```
Comment (maleck13) [inline src/main/java/org/aerogear/plugin/intellij/mobile/module/AppRepresentationModuleBuilder.java]: Do you have to override these?

File: src/main/java/org/aerogear/plugin/intellij/mobile/module/RsSdkComboBox.java [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+import com.intellij.ui.ComboboxWithBrowseButton;
+import com.intellij.ui.SimpleTextAttributes;
+
+import javax.swing.*;
+import java.awt.event.ActionEvent;
+import java.awt.event.ActionListener;
+import java.util.List;
+
+/**
+ * @author Konstantin Bulenkov
```
Comment (maleck13) [inline src/main/java/org/aerogear/plugin/intellij/mobile/module/RsSdkComboBox.java]: who?

File: src/main/java/org/aerogear/plugin/intellij/mobile/module/RsSdkPanel.java [hunk 4]
```diff
@@ -0,0 +1,103 @@
...
+              AccessToken accessToken = ApplicationManager.getApplication().acquireWriteActionLock(RsSdkPanel.class);
+              try {
+                if (!dirName.toLowerCase().contains("smalltalk") && !dirName.toLowerCase().contains("redline")) {
+                try {
+                  dir = dir.createChildDirectory(this, "RedlineSmalltalk");
+                } catch (IOException e) {//
+                }
+              }
+              DownloadableFileService fileService = DownloadableFileService.getInstance();
+              DownloadableFileDescription fileDescription = fileService.createFileDescription("https://github.com/redline-smalltalk/redline-smalltalk.github.com/raw/master/assets/redline-deploy.zip", "redline-deploy.zip");
```
Comment (maleck13) [inline src/main/java/org/aerogear/plugin/intellij/mobile/module/RsSdkPanel.java]: not sure what this is

Comment (witmicko) [reply src/main/java/org/aerogear/plugin/intellij/mobile/module/RsSdkPanel.java]: @maleck13 copied over from their examples so we could work on it in parallel, it was early stages yesterday.

Comment (maleck13) [reply src/main/java/org/aerogear/plugin/intellij/mobile/module/RsSdkPanel.java]: ok sorry :)

File: src/main/java/org/aerogear/plugin/intellij/mobile/module/AppRepresentationModuleType.java [hunk 5]
```diff
@@ -0,0 +1,57 @@
...
+  @Override
+  public Icon getNodeIcon(boolean isOpened) {
+    return MobileIcons.Feedhenry;
+  }
+
+//  @NotNull
+//  @Override
+//  public ModuleWizardStep[] createWizardSteps(@NotNull WizardContext wizardContext, @NotNull AppRepresentationModuleBuilder moduleBuilder, @NotNull ModulesProvider modulesProvider) {
+//    return super.createWizardSteps(wizardContext, moduleBuilder, modulesProvider);
+//  }
```
Comment (secondsun) [inline src/main/java/org/aerogear/plugin/intellij/mobile/module/AppRepresentationModuleType.java]: commented out code should be removed

File: src/main/java/org/aerogear/plugin/intellij/mobile/module/CreateClientForm.java [hunk 6]
```diff
@@ -0,0 +1,71 @@
...
+    //TODO add available namespaces dynamically
+    namespaceComboBox = new ComboBox(new String[]{"myproject"});
+  }
+
+  /**
+   * Returns form as map <name, value>
+   *
+   * @return map
+   */
+  public Map getInputs() {
```
Comment (secondsun) [inline src/main/java/org/aerogear/plugin/intellij/mobile/module/CreateClientForm.java]: Needs generic arguments

File: src/main/java/org/aerogear/plugin/intellij/mobile/module/CreateClientForm.java [hunk 7]
```diff
@@ -0,0 +1,71 @@
...
+    return map;
+  }
+
+  /**
+   * Set border RED on clientName input to indicate missing/invalid input.
+   */
+  public void invalidNameNotify() {
+    errorLabel.setText("Error");
+    errorMessage.setText("Client Name is required");
+    defaultClientNameBorder =     clientNameTxtField.getBorder();
```
Comment (secondsun) [inline src/main/java/org/aerogear/plugin/intellij/mobile/module/CreateClientForm.java]: extra white space

File: src/main/java/org/aerogear/plugin/intellij/mobile/ui/MobileIcons.java [hunk 8]
```diff
@@ -0,0 +1,9 @@
+package org.aerogear.plugin.intellij.mobile.ui;
+
+import com.intellij.openapi.util.IconLoader;
+
+import javax.swing.*;
+
+public class MobileIcons {
+  public static final Icon Feedhenry = IconLoader.get1xIcon(IconLoader.getIcon("/icons/fh_icon.png"));
```
Comment (secondsun) [inline src/main/java/org/aerogear/plugin/intellij/mobile/ui/MobileIcons.java]: Feedhenry should be in all caps.

File: src/test/java/org/aerogear/plugin/intellij/mobile/api/MobileAPITest.java [hunk 9]
```diff
@@ -0,0 +1,14 @@
...
+public class MobileAPITest {
+    @BeforeClass
+    public void setup() {
+    }
+
+    @Test
+    public void testGetServices() {
+    }
+
+}
```
Comment (secondsun) [inline src/test/java/org/aerogear/plugin/intellij/mobile/api/MobileAPITest.java]: remove empty test class

File: gradle/wrapper/gradle-wrapper.properties [hunk 10]
```diff
@@ -1,6 +1,5 @@
-#Fri Jan 12 14:51:45 GMT 2018
 distributionBase=GRADLE_USER_HOME
 distributionPath=wrapper/dists
 zipStoreBase=GRADLE_USER_HOME
 zipStorePath=wrapper/dists
-distributionUrl=https\://services.gradle.org/distributions/gradle-3.5-rc-2-all.zip
+distributionUrl=https\://services.gradle.org/distributions/gradle-3.0-bin.zip
```
Comment (secondsun) [inline gradle/wrapper/gradle-wrapper.properties]: This is the incorrect gradle version to use.  Please use 4.1-all

File: src/main/java/org/aerogear/plugin/intellij/mobile/constants/Constants.java [hunk 11]
```diff
@@ -0,0 +1,9 @@
+package org.aerogear.plugin.intellij.mobile.constants;
+
+public class Constants {
+  public static final  String[] CLIENT_TYPES = new String[]{
```
Comment (secondsun) [inline src/main/java/org/aerogear/plugin/intellij/mobile/constants/Constants.java]: extra whitespace before String[]

File: src/main/java/org/aerogear/plugin/intellij/mobile/module/AppRepresentationModuleBuilder.java [hunk 12]
```diff
@@ -0,0 +1,45 @@
...
+import org.jetbrains.annotations.NotNull;
+import org.jetbrains.annotations.Nullable;
+
+import javax.swing.*;
+
+public class AppRepresentationModuleBuilder extends ModuleBuilder implements ModuleBuilderListener{
+
+  @Override
+  public void setupRootModel(ModifiableRootModel modifiableRootModel) throws ConfigurationException {
+    //TODO
```
Comment (secondsun) [inline src/main/java/org/aerogear/plugin/intellij/mobile/module/AppRepresentationModuleBuilder.java]: Add some description to the todo block

File: src/main/java/org/aerogear/plugin/intellij/mobile/module/AppRepresentationModuleBuilder.java [hunk 13]
```diff
@@ -0,0 +1,45 @@
...
+  @Nullable
+  @Override
+  public ModuleWizardStep getCustomOptionsStep(WizardContext context, Disposable parentDisposable) {
+    return new AppRepresentationFirstStepWizard();
+  }
+
+
+  @Override
+  public void moduleCreated(@NotNull Module module) {
+    //TODO
```
Comment (secondsun) [inline src/main/java/org/aerogear/plugin/intellij/mobile/module/AppRepresentationModuleBuilder.java]: Add some description to the todo block

File: src/main/java/org/aerogear/plugin/intellij/mobile/module/AppRepresentationModuleType.java [hunk 14]
```diff
@@ -0,0 +1,57 @@
...
+import com.intellij.ide.util.projectWizard.ModuleWizardStep;
+import com.intellij.ide.util.projectWizard.WizardContext;
+import com.intellij.openapi.module.ModuleType;
+import com.intellij.ide.util.projectWizard.EmptyModuleBuilder;
+import com.intellij.openapi.module.ModuleTypeManager;
+import com.intellij.openapi.roots.ui.configuration.ModulesProvider;
+import org.aerogear.plugin.intellij.mobile.ui.MobileIcons;
+import org.jetbrains.annotations.NotNull;
+
+import javax.swing.*;
```
Comment (secondsun) [inline src/main/java/org/aerogear/plugin/intellij/mobile/module/AppRepresentationModuleType.java]: avoid wildcard imports

File: src/main/java/org/aerogear/plugin/intellij/mobile/ui/MobileIcons.java [hunk 15]
```diff
@@ -0,0 +1,9 @@
+package org.aerogear.plugin.intellij.mobile.ui;
+
+import com.intellij.openapi.util.IconLoader;
+
+import javax.swing.*;
+
+public class MobileIcons {
+  public static final Icon FEEDHENRY = IconLoader.get1xIcon(IconLoader.getIcon("/icons/fh_icon.png"));
```
Comment (secondsun) [inline src/main/java/org/aerogear/plugin/intellij/mobile/ui/MobileIcons.java]: Can we get the AeroGear icon?

Comment (witmicko) [reply src/main/java/org/aerogear/plugin/intellij/mobile/ui/MobileIcons.java]: @secondsun thank you for reviews so far, can you give it another look, please. it is not WIP anymore :)

File: src/main/java/org/aerogear/plugin/intellij/mobile/wizard/Constants.java [hunk 16]
```diff
@@ -0,0 +1,13 @@
+package org.aerogear.plugin.intellij.mobile.wizard;
+
+class Constants {
+    public static final String[] CLIENT_TYPES = new String[]{
+            "android",
+            "cordova",
+            "iOS"
+    };
+
+    public static final String MODULE_ID = "MOBILE_MODULE";
```
Comment (vchepeli) [inline src/main/java/org/aerogear/plugin/intellij/mobile/wizard/Constants.java]: replace MODULE_ with CLIENT_APP

File: src/main/java/org/aerogear/plugin/intellij/mobile/module/RsSdkComboBox.java (hunk 3 above)
Comment (vchepeli) [reply src/main/java/org/aerogear/plugin/intellij/mobile/module/RsSdkComboBox.java]: @witmicko now you are russian :D

File: src/main/java/org/aerogear/plugin/intellij/mobile/ui/MobileIcons.java (hunk 15 above)
Comment (secondsun) [reply src/main/java/org/aerogear/plugin/intellij/mobile/ui/MobileIcons.java]: Sure if you will give https://github.com/aerogear/aerogear-mobile-intellij-plugin/pull/10 a look.

Comment (secondsun) [reply src/main/java/org/aerogear/plugin/intellij/mobile/ui/MobileIcons.java]: Hrm, the javadocs for the `get1xIcon` say " For internal usage. Converts the icon to 1x scale when applicable.".  I don't think that we should be using this method.  Furthermore this API isn't available in Android Studio which is a pretty big deal breaker.

File: src/main/java/org/aerogear/plugin/intellij/mobile/wizard/CreateClientForm.java [hunk 17]
```diff
@@ -0,0 +1,142 @@
...
+    private JTextField clientAppIdTxtField;
+    private JLabel errorLabel;
+    private JLabel errorMessage;
+    private Border defaultClientNameBorder;
+    private CreateClientFormInputs formInputs;
+
+
+    CreateClientForm() {
+        super();
+        $$$setupUI$$$();
```
Comment (secondsun) [inline src/main/java/org/aerogear/plugin/intellij/mobile/wizard/CreateClientForm.java]: I think that it might be great for our sanity if we put tool generated code in its own namespace and create a subclass for our APIs that we ultimately use.  I don't have that much experience with UI code like this in large projects, but that is my gut instinct.

What I am trying to avoid is a situation where we can't use the GUI editor and have to rewrite a lot of code.

File: src/main/java/org/aerogear/plugin/intellij/mobile/wizard/CreateClientFormInputs.java [hunk 18]
```diff
@@ -0,0 +1,36 @@
+package org.aerogear.plugin.intellij.mobile.wizard;
+
+public class CreateClientFormInputs {
+    private String name;
+    private String clientType;
+    private String appIdentifier;
+
+
+    CreateClientFormInputs() {
```
Comment (secondsun) [inline src/main/java/org/aerogear/plugin/intellij/mobile/wizard/CreateClientFormInputs.java]: Should this constructor become public?

File: src/main/java/org/aerogear/plugin/intellij/mobile/wizard/CreateClientFormInputs.java [hunk 19]
```diff
@@ -0,0 +1,36 @@
+package org.aerogear.plugin.intellij.mobile.wizard;
+
+public class CreateClientFormInputs {
```
Comment (secondsun) [inline src/main/java/org/aerogear/plugin/intellij/mobile/wizard/CreateClientFormInputs.java]: I think the class and its fields need javadocs.

File: src/main/java/org/aerogear/plugin/intellij/mobile/wizard/CreateClientFirstStep.java [hunk 20]
```diff
@@ -0,0 +1,66 @@
...
+            output = mobileClient.getSpec().toString();
+
+        } catch (CLIException e) {
+            output = e.getMessage();
+        }
+        createClientBuilder.setCommandOutput(output);
+        System.out.println(output);
+    }
+
+    public boolean validate() {
```
Comment (secondsun) [inline src/main/java/org/aerogear/plugin/intellij/mobile/wizard/CreateClientFirstStep.java]: Does this need an `@override`?

File: src/main/java/org/aerogear/plugin/intellij/mobile/wizard/CreateClientBuilder.java [hunk 21]
```diff
@@ -0,0 +1,55 @@
...
+        //TODO what do we do after the wizard has been created
+    }
+
+
+    @Override
+    public ModuleWizardStep[] createWizardSteps(@NotNull WizardContext wizardContext, @NotNull ModulesProvider modulesProvider) {
+        return new ModuleWizardStep[]{new CreateClientSecondStep(this)};
+    }
+
+    public String getCommandOutput() {
```
Comment (secondsun) [inline src/main/java/org/aerogear/plugin/intellij/mobile/wizard/CreateClientBuilder.java]: Javadocs for this property.

File: src/main/java/org/aerogear/plugin/intellij/mobile/wizard/CreateClientSecondStep.java [hunk 22]
```diff
@@ -0,0 +1,42 @@
...
+        textArea.setText(myBuilder.getCommandOutput());
+        return textArea;
+    }
+
+    @Override
+    public void updateDataModel() {
+
+    }
+
+    public boolean validate() {
```
Comment (secondsun) [inline src/main/java/org/aerogear/plugin/intellij/mobile/wizard/CreateClientSecondStep.java]: Does this need @Override?

File: src/main/java/org/aerogear/plugin/intellij/mobile/wizard/CreateClientFormInputs.java (hunk 18 above)
Comment (witmicko) [reply src/main/java/org/aerogear/plugin/intellij/mobile/wizard/CreateClientFormInputs.java]: don't think so, it is only ever used within `wizard` package so package private is fine in my opinion

File: src/main/java/org/aerogear/plugin/intellij/mobile/wizard/CreateClientForm.java (hunk 17 above)
Comment (witmicko) [reply src/main/java/org/aerogear/plugin/intellij/mobile/wizard/CreateClientForm.java]: totaly agree, 'fixed' that by changing this option in the IDE
image

File: src/main/java/org/aerogear/plugin/intellij/mobile/api/MobileAPI.java [hunk 23]
```diff
@@ -42,25 +41,21 @@ public void createService(ServiceClass sc, List<String> params, Watch w) {
...
         cliRunner.executeAndWatch(cmd, w);
     }
-    
-    
-    public MobileClient createClient(String name, String clientType, String bundleID)throws CLIException{
-        if("".equals(name) || "".equals(clientType) || "".equals(bundleID)){
+
+
+    public MobileClient createClient(String name, String clientType, String bundleID) throws CLIException {
+        if ("".equals(name) || "".equals(clientType) || "".equals(bundleID)) {
```
Comment (dimitraz) [inline src/main/java/org/aerogear/plugin/intellij/mobile/api/MobileAPI.java]: maybe change these to `isEmpty()`

File: src/main/java/org/aerogear/plugin/intellij/mobile/wizard/CreateClientFirstStep.java [hunk 24]
```diff
@@ -0,0 +1,66 @@
+package org.aerogear.plugin.intellij.mobile.wizard;
+
+import com.intellij.ide.util.projectWizard.ModuleWizardStep;
+import org.aerogear.plugin.intellij.mobile.api.CLIException;
+import org.aerogear.plugin.intellij.mobile.api.CliRunnerImpl;
```
Comment (dimitraz) [inline src/main/java/org/aerogear/plugin/intellij/mobile/wizard/CreateClientFirstStep.java]: can you rename this class to CLIRunnerImpl

File: src/main/java/org/aerogear/plugin/intellij/mobile/ui/MobileIcons.java [hunk 25]
```diff
@@ -0,0 +1,10 @@
+package org.aerogear.plugin.intellij.mobile.ui;
+
+import com.intellij.openapi.util.IconLoader;
+
+import javax.swing.*;
+
+public class MobileIcons {
+  public static final Icon FEEDHENRY = IconLoader.getIcon("/icons/fh_icon.png");
```
[the remark this answers is not in the document]
Comment (witmicko) [reply src/main/java/org/aerogear/plugin/intellij/mobile/ui/MobileIcons.java]: Not quite sure about this, I changed it from `get1x` which is internal to `getIcon()`
https://github.com/JetBrains/intellij-community/blob/master/platform/util/src/com/intellij/openapi/util/IconLoader.java#L387
https://github.com/JetBrains/intellij-community/blob/master/platform/util/src/com/intellij/openapi/util/IconLoader.java#L102

intelij.openapi seems to be compatible with all Intelij products:
https://www.jetbrains.org/intellij/sdk/docs/basics/getting_started/plugin_compatibility.html

@secondsun can you please comment on this, plugin was tested in Android Studio and Webstorm.

**Comment by wtrocki** [comment]
Creating mobile client app is less flexible for developers than just using file wizard (mobile client file). 
Are we doing anything on the project level?

**Comment by witmicko** [comment]
> Creating mobile client app is less flexible for developers than just using file wizard (mobile client file).
agree
> Are we doing anything on the project level?
not at this moment, this approach seems to work in IDEA but project wizard - plugin integration seems to be more restricted in other Intelij products.
